packages/http/fetch/src/middlewares/compressionHandler.ts

```diff
@@ -0,0 +1,239 @@
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
+ * See License in the project root for license information.
+ * -------------------------------------------------------------------------------------------
+ */
+
+import { type RequestOption, inNodeEnv } from "@microsoft/kiota-abstractions";
+import { Span, trace } from "@opentelemetry/api";
+
+import { getObservabilityOptionsFromRequest } from "../observabilityOptions";
+import type { Middleware } from "./middleware";
+import { CompressionHandlerOptions, CompressionHandlerOptionsKey } from "./options/compressionHandlerOptions";
+import type { FetchHeadersInit, FetchRequestInit } from "../utils/fetchDefinitions";
+import { deleteRequestHeader, getRequestHeader, setRequestHeader } from "../utils/headersUtil";
+
+/**
+ * Compress the url content.
+ */
+export class CompressionHandler implements Middleware {
+	next: Middleware | undefined;
+
+	/**
+	 * @private
+	 * @static
+	 * A member holding the name of content range header
+	 */
+	private static readonly CONTENT_RANGE_HEADER = "Content-Range";
+
+	/**
+	 * @private
+	 * @static
+	 * A member holding the name of content encoding header
+	 */
+	private static readonly CONTENT_ENCODING_HEADER = "Content-Encoding";
+
+	/**
+	 * @public
+	 * @constructor
+	 * Creates a new instance of the CompressionHandler class
+	 * @param {CompressionHandlerOptions} handlerOptions The options for the compression handler.
+	 * @returns An instance of the CompressionHandler class
+	 */
+	public constructor(private readonly handlerOptions: CompressionHandlerOptions = new CompressionHandlerOptions()) {
+		if (!handlerOptions) {
+			throw new Error("handlerOptions cannot be undefined");
+		}
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public execute(url: string, requestInit: RequestInit, requestOptions?: Record<string, RequestOption> | undefined): Promise<Response> {
+		let currentOptions = this.handlerOptions;
+		if (requestOptions?.[CompressionHandlerOptionsKey]) {
+			currentOptions = requestOptions[CompressionHandlerOptionsKey] as CompressionHandlerOptions;
+		}
+		const obsOptions = getObservabilityOptionsFromRequest(requestOptions);
+		if (obsOptions) {
+			return trace.getTracer(obsOptions.getTracerInstrumentationName()).startActiveSpan("compressionHandler - execute", (span) => {
+				try {
+					span.setAttribute("com.microsoft.kiota.handler.compression.enable", currentOptions.ShouldCompress);
+					return this.executeInternal(currentOptions, url, requestInit as FetchRequestInit, requestOptions, span);
+				} finally {
+					span.end();
+				}
+			});
+		}
+		return this.executeInternal(currentOptions, url, requestInit as FetchRequestInit, requestOptions);
+	}
+
+	private async executeInternal(options: CompressionHandlerOptions, url: string, requestInit: FetchRequestInit, requestOptions?: Record<string, RequestOption> | undefined, span?: Span): Promise<Response> {
```

Comment:
```suggestion
	private async executeInternal(options: CompressionHandlerOptions, url: string, requestInit: FetchRequestInit, requestOptions?: Record<string, RequestOption>, span?: Span): Promise<Response> {
```
parameter is already optional..